### DISCUSSION TITLE
Can the viewer show 2D primitives?

### DISCUSSION THREAD

**Discussion by t1u1** [jscad/OpenJSCAD.org#987] [Q&A] [answered]
Hi,

I am using OpenJSCAD programmatically. It works fine 90% of the time, but there seem to be some bugs in edge cases. In order to report the bugs, I am trying to minimise the test cases. And for that, I am trying to remove a few steps from the code.

My current code is along these lines:
1. Create 2D geometry using `primitives.circle()` etc
2. Extrude this to 3D using `extrusions.extrudeLinear()`
3. Display this using `entitiesFromSolids()`

I want to eliminate step 2, since a known bug is present in extrusion (https://github.com/jscad/OpenJSCAD.org/issues/907)

If I don't do step 2 (extrusion), the viewer doesn't show anything on screen. 

Is it possible to show the 2D geometry directly?

**Comment by z3dev** [comment, marked as answer]
@t1u1 welcome

i wouldn't worry about #907 as that's a very complex edge case. you probably won't have any issues with extrudeLinear()

the simplest viewer can be found in the regl_renderer directory, called demo.hmtl

that should get you going

**Reply by t1u1** [reply]
My use case also looks complex. Perhaps even more complex than the letter 7. See this screenshot (when it's working):

![working](https://user-images.githubusercontent.com/90307689/152309958-a084d722-38ac-4282-aaec-4f6619ed8a7f.png)

And when it's not working, when the model dimension is slightly altered:

![notWorking](https://user-images.githubusercontent.com/90307689/152310052-b7a78988-cb78-4613-a360-aa265d2abd23.png)

In fact, I am now very sure that it's extrusion problem, because when I convert the exact same 2D geometry to SVG, it works fine.

----

I based my renderer on `demo.html` itself. If I just pass 2D geometry to it, it shows nothing. Am I supposed to call something other than `entitiesFromSolids()` for 2D geom?

**Reply by z3dev** [reply]
We would love to have your design for testing. Please create a new issue if you can make a small test case.

**Reply by t1u1** [reply]
I have been trying to think how I could make a simpler test case. The test of letter 7 seems to be very simple. If that itself is hard to solve, any part of my model which shows the bug is way more complex than that! :laughing:

**Comment by hrgdavor** [comment]
@t1u1 yes you can. just return that 2d primitive from main

```js
const jscad = require('@jscad/modeling')
const { circle } = jscad.primitives

const main = ()=>{
  return circle({radius: 15})
}

module.exports = {main}
```

![image](https://user-images.githubusercontent.com/2480762/152304058-c26222bb-18fd-4c86-a102-e66c79fa4d22.png)

**Reply by t1u1** [reply]
That's good to hear. But I haven't been able to use the web editor at https://openjscad.xyz/
After typing the code it never seems to update. What is the trigger for updating the view after typing the code? I tried "Ctrl+Enter" or closing and opening the editor, or clicking elsewhere, but nothing works. And there is no message in the Debug console.

Hence, I created my own renderer based on demo.html. But in this I am not able to render 2D geom.

**Reply by z3dev** [reply]
There's some help on the key strokes. Sorry that you didn't find those.

It's easier to use your favorite editor, and just drag and drop the file to the web page. You can also set the "Enable auto-reload", and that will reload the design whenever you make changes. Cool stuff.

**Reply by hrgdavor** [reply]
![image](https://user-images.githubusercontent.com/2480762/152313400-9098612c-7eec-42ae-b92c-f2ce3ddd6e61.png)

I would also recomment using your own editor, saving file locally and using drag and drop to view the result and use the "Enable auto-reload"

**Reply by t1u1** [reply]
Ah, it's `shift+enter`, thanks! It is now working. Should have looked more closely :smile:

**Comment by t1u1** [comment]
OK, I figured out what was missing in my renderer: the drawLines command!

Rendering of 2D primitives is now working! Thanks for the help @hrgdavor @z3dev 

I can now try to isolate the boolean bugs from the extrusion bugs.
